### PR TITLE
CB-4705 make data viewer model loaders more flexible

### PR DESCRIPTION
**Pull request by devnaumov**

### CODE REVIEW

File: webapp/packages/plugin-data-viewer/src/useDataViewerModel.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        return this._exception;
+      },
+      isLoading(): boolean {
+        return connection.isLoading() || this._loading;
+      },
+      isLoaded(): boolean {
+        return connection.isLoaded() && this.dataViewerTable.get(this.tab.handlerState.tableId || '') !== undefined;
+      },
+      isError(): boolean {
+        return this.exception !== null;
```
Comment (sergeyteleshev) [inline webapp/packages/plugin-data-viewer/src/useDataViewerModel.ts]: exception is also can be undefined

I believe something like this is more suitable here:
`return isNotNullDefined(this.exception)`

File: webapp/packages/plugin-data-viewer/src/ContainerDataSource.ts [hunk 2]
```diff
@@ -233,7 +250,7 @@ export class ContainerDataSource extends ResultSetDataSource<IDataContainerOptio
     }));
   }
 
-  private async ensureContextCreated(): Promise<IConnectionExecutionContext> {
+  private async getContext(): Promise<IConnectionExecutionContext> {
```
Comment (Wroud) [inline webapp/packages/plugin-data-viewer/src/ContainerDataSource.ts]: please revert, this is not obvious that get* function can create context

File: webapp/packages/plugin-data-viewer/src/DataViewerTable.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import type { Connection } from '@cloudbeaver/core-connections';
+
+import type { IDatabaseDataModel } from './DatabaseDataModel/IDatabaseDataModel';
+import type { IDatabaseResultSet } from './DatabaseDataModel/IDatabaseResultSet';
+import type { IDataViewerTableStorage } from './IDataViewerTableStorage';
+
+export abstract class DataViewerTable {
```
Comment (Wroud) [inline webapp/packages/plugin-data-viewer/src/DataViewerTable.ts]: should not exists

File: webapp/packages/plugin-data-viewer/src/DataViewerTableService.ts [hunk 4]
```diff
@@ -17,10 +17,11 @@ import { DatabaseDataAccessMode } from './DatabaseDataModel/IDatabaseDataSource'
 import type { IDatabaseResultSet } from './DatabaseDataModel/IDatabaseResultSet';
 import { DataViewerService } from './DataViewerService';
 import { DataViewerSettingsService } from './DataViewerSettingsService';
+import { DataViewerTable } from './DataViewerTable';
 import { TableViewerStorageService } from './TableViewer/TableViewerStorageService';
 
 @injectable()
-export class DataViewerTableService {
+export class DataViewerTableService extends DataViewerTable {
```
Comment (Wroud) [inline webapp/packages/plugin-data-viewer/src/DataViewerTableService.ts]: please revert

**Review by sergeyteleshev — APPROVED** [review]

File: webapp/packages/plugin-data-viewer/src/useDataViewerModel.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  dataViewerTable: DataViewerTable;
+  tab: Tab;
+  _exception?: Error[] | Error | null;
+  _loading: boolean;
+  _init(): Promise<void>;
+  init(tab: Tab, state: State): Promise<void>;
+  load(): Promise<void>;
+}
+
+export function useDataViewerModel<T extends Tab>(tab: T, dataViewerTable: DataViewerTable, init: (tab: T, state: State) => Promise<void>) {
```
Comment (Wroud) [inline webapp/packages/plugin-data-viewer/src/useDataViewerModel.ts]: DatabaseDataModel should not be associated with Tabs, please pass needed information as arguments to this hook, you don't need DataViewerTable here, you can use `TableViewerStorageService` directly

to init function you can pass DatabaseDataModel and probably you don't need to pass state

File: webapp/packages/plugin-data-viewer/src/DataViewerTableService.ts [hunk 6]
```diff
@@ -30,23 +31,8 @@ export class DataViewerTableService {
...
-  }
-
-  async removeTableModel(tableId: string): Promise<void> {
-    const model = this.tableViewerStorageService.get(tableId);
-
-    if (model) {
-      this.tableViewerStorageService.remove(tableId);
-      await model.dispose();
-    }
+  ) {
```
Comment (Wroud) [inline webapp/packages/plugin-data-viewer/src/DataViewerTableService.ts]: seems like we must remove this methods from `DataViewerTableService` and `CollectionViewerTableService` (and maybe other places) and use `TableViewerStorageService` instead

File: webapp/packages/plugin-data-viewer/src/DataViewerPage/useDataViewerPanel.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  const model = useDataViewerModel(objectViewerTab, dataViewerTableService, async (tab, state) => {
+    const node = navNodeManagerService.getNode({
+      nodeId: tab.handlerState.objectId,
+      parentId: tab.handlerState.parentId,
+    });
+
+    if (!navNodeManagerService.isNodeHasData(node)) {
+      state.resetException();
+      return;
+    }
```
Comment (Wroud) [inline webapp/packages/plugin-data-viewer/src/DataViewerPage/useDataViewerPanel.ts]: should be handled by useDataViewerModel. Don't call `state.resetException();` in such contexts (if init not throw an exception, then it should be cleared by `useDataViewerModel`)

File: webapp/packages/plugin-data-viewer/src/ContainerDataSource.ts [hunk 8]
```diff
@@ -217,6 +182,58 @@ export class ContainerDataSource extends ResultSetDataSource<IDataContainerOptio
     return prevResults;
   }
 
+  getResultId(prevResults: IDatabaseResultSet[], context: IConnectionExecutionContextInfo) {
```
Comment (Wroud) [inline webapp/packages/plugin-data-viewer/src/ContainerDataSource.ts]: They all public what is the reason for it?

**Review by Wroud — CHANGES_REQUESTED** [review]

File: webapp/packages/plugin-data-viewer/src/useDataViewerModel.ts (hunk 1 above)
Comment (Wroud) [reply webapp/packages/plugin-data-viewer/src/useDataViewerModel.ts]: right, but as long as it implements ILoadableState it's better to use `isContainsException` cuz it accepts all variants of exception value

File: webapp/packages/plugin-data-viewer/src/DataViewerPage/DataViewerPanel.tsx [hunk 9]
```diff
@@ -8,16 +8,17 @@
 import { observer } from 'mobx-react-lite';
 import { useCallback } from 'react';
 
-import { Loader, TextPlaceholder } from '@cloudbeaver/core-blocks';
+import { Loader, TextPlaceholder, useAutoLoad, useTranslate } from '@cloudbeaver/core-blocks';
```
Comment (sergeyteleshev) [inline webapp/packages/plugin-data-viewer/src/DataViewerPage/DataViewerPanel.tsx]: pls remove unused loader here

File: webapp/packages/plugin-data-viewer/src/DataViewerPage/useDataViewerPanel.ts [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        await connectionInfoResource.waitLoad();
+        const connectionInfo = connectionInfoResource.get(tab.handlerState.connectionKey!);
+
+        if (!connectionInfo) {
+          throw new Error("Connection doesn't exists");
+        }
+
+        model = dataViewerTableService.create(connectionInfo, node);
+        tab.handlerState.tableId = model.id;
+        model.source.setOutdated();
```
Comment (sergeyteleshev) [inline webapp/packages/plugin-data-viewer/src/DataViewerPage/useDataViewerPanel.ts:58]: why do we need it to set it to outdated here? may be we should incapsulate this logic inside 
`dataViewerTableService.create()` function?

**Review by sergeyteleshev — APPROVED** [review]

**Review by Wroud — APPROVED** [review]

File: webapp/packages/plugin-data-viewer/src/DataViewerPage/useDataViewerPanel.ts (hunk 10 above)
Comment (devnaumov) [reply webapp/packages/plugin-data-viewer/src/DataViewerPage/useDataViewerPanel.ts:58]: I think its better when create fn just create the model without any side effects

**Pull request merged**